Non-reactive forceFlush in AbstractReactiveSaveEventListener.java:172

[There](https://github.com/hibernate/hibernate-reactive/blob/main/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/AbstractReactiveSaveEventListener.java#L172) is a call to non-reactive forceFlush of `SessionImpl` that leads to UnsupportedOperationException from DefaultReactiveFlushEventListener:

```
	Caused by: java.util.concurrent.CompletionException: javax.persistence.PersistenceException: org.hibernate.HibernateException: java.lang.UnsupportedOperationException
		at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)
		at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1194)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
		at org.hibernate.reactive.event.impl.AbstractReactiveSaveEventListener.reactiveSaveWithGeneratedId(AbstractReactiveSaveEventListener.java:123)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.saveTransientEntity(DefaultReactiveMergeEventListener.java:283)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.lambda$entityIsTransient$5(DefaultReactiveMergeEventListener.java:257)
		at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.entityIsTransient(DefaultReactiveMergeEventListener.java:257)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.lambda$entityIsDetached$11(DefaultReactiveMergeEventListener.java:364)
		at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.entityIsDetached(DefaultReactiveMergeEventListener.java:321)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.reactiveOnMerge(DefaultReactiveMergeEventListener.java:190)
		at org.hibernate.reactive.event.impl.DefaultReactiveMergeEventListener.reactiveOnMerge(DefaultReactiveMergeEventListener.java:90)
		at org.hibernate.event.service.internal.EventListenerGroupImpl.lambda$fireEventOnEachListener$0(EventListenerGroupImpl.java:133)
		at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2309)
		at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:144)
		at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:133)
		at org.hibernate.reactive.session.impl.ReactiveSessionImpl.fireMerge(ReactiveSessionImpl.java:849)
		... 298 more
	Caused by: javax.persistence.PersistenceException: org.hibernate.HibernateException: java.lang.UnsupportedOperationException
		at app//org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:154)
		at app//org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
		at app//org.hibernate.reactive.session.impl.ReactiveExceptionConverter.convert(ReactiveExceptionConverter.java:31)
		at app//org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1411)
		at app//org.hibernate.internal.SessionImpl.forceFlush(SessionImpl.java:1438)
		at app//org.hibernate.reactive.event.impl.AbstractReactiveSaveEventListener.reactivePerformSave(AbstractReactiveSaveEventListener.java:172)
		at app//org.hibernate.reactive.event.impl.AbstractReactiveSaveEventListener.lambda$reactiveSaveWithGeneratedId$0(AbstractReactiveSaveEventListener.java:123)
		at java.base@17.0.1/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1187)
		... 320 more
	Caused by: org.hibernate.HibernateException: java.lang.UnsupportedOperationException
		... 326 more
	Caused by: java.lang.UnsupportedOperationException
		at org.hibernate.reactive.event.impl.DefaultReactiveFlushEventListener.onFlush(DefaultReactiveFlushEventListener.java:63)
		at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
		at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1407)
		... 324 more
```

I ran your in test with an ORM test in hibernate-core

```
Caused by: org.hibernate.PersistentObjectException: detached entity passed to persist: org.hibernate.test.any.Person
	at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:120)
	at org.hibernate.event.internal.DefaultPersistEventListener.onPersist(DefaultPersistEventListener.java:55)
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:107)
	at org.hibernate.internal.SessionImpl.firePersist(SessionImpl.java:756)
	... 17 more
```

Your test ends up with something similar...

```
Caused by: org.hibernate.PersistentObjectException: detached entity passed to persist: org.hibernate.reactive.ORMForceFlushTest$Flour
	at app//org.hibernate.reactive.event.impl.DefaultReactivePersistEventListener.reactiveOnPersist(DefaultReactivePersistEventListener.java:121)
	at app//org.hibernate.reactive.event.impl.DefaultReactivePersistEventListener.reactiveOnPersist(DefaultReactivePersistEventListener.java:58)
	at app//org.hibernate.event.service.internal.EventListenerGroupImpl.lambda$fireEventOnEachListener$0(EventListenerGroupImpl.java:133)
	at org.hibernate.reactive.session.impl.ReactiveSessionImpl.firePersist(ReactiveSessionImpl.java:699)
```
The user's exception is thrown from Reactive is thrown from the same ORM class: `EventListenerGroupImpl.fireEventOnEachListener()' but triggered via `doFlush()`

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/AbstractReactiveSaveEventListener.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/event/impl/AbstractReactiveSaveEventListener.java
@@ -35,6 +35,7 @@ import org.hibernate.reactive.engine.impl.ReactiveEntityRegularInsertAction;
 import org.hibernate.reactive.logging.impl.Log;
 import org.hibernate.reactive.logging.impl.LoggerFactory;
 import org.hibernate.reactive.session.ReactiveSession;
+import org.hibernate.reactive.util.impl.CompletionStages;
 import org.hibernate.type.Type;
 import org.hibernate.type.TypeHelper;
 
@@ -162,26 +163,14 @@ abstract class AbstractReactiveSaveEventListener<C> implements CallbackRegistryC
 			LOG.tracev( "Saving {0}", infoString( persister, id, source.getFactory() ) );
 		}
 
-		final EntityKey key;
-		if ( !useIdentityColumn ) {
-			key = source.generateEntityKey( id, persister );
-			final PersistenceContext persistenceContext = source.getPersistenceContextInternal();
-			Object old = persistenceContext.getEntity( key );
-			if ( old != null ) {
-				if ( persistenceContext.getEntry( old ).getStatus() == Status.DELETED ) {
-					source.forceFlush( persistenceContext.getEntry( old ) );
-				}
-				else {
-					return failedFuture( new NonUniqueObjectException( id, persister.getEntityName() ) );
-				}
-			}
+		CompletionStage<EntityKey> keyStage = useIdentityColumn
+				? CompletionStages.nullFuture()
+				: generateEntityKey( id, persister, source ).thenCompose( generatedKey -> {
 			persister.setIdentifier( entity, id, source );
-		}
-		else {
-			key = null;
-		}
+			return CompletionStages.completedFuture( generatedKey );
+		} );
 
-		return reactivePerformSaveOrReplicate(
+		return keyStage.thenCompose( key -> reactivePerformSaveOrReplicate(
 				entity,
 				key,
 				persister,
@@ -189,7 +178,25 @@ abstract class AbstractReactiveSaveEventListener<C> implements CallbackRegistryC
 				context,
 				source,
 				requiresImmediateIdAccess
-		);
+		) );
+	}
+
+	private CompletionStage<EntityKey> generateEntityKey(Serializable id, EntityPersister persister,
+			EventSource source) {
+		final EntityKey key = source.generateEntityKey( id, persister );
+		final PersistenceContext persistenceContext = source.getPersistenceContextInternal();
+		Object old = persistenceContext.getEntity( key );
+		if ( old != null ) {
+			if ( persistenceContext.getEntry( old ).getStatus() == Status.DELETED ) {
+				return source.unwrap( ReactiveSession.class ).reactiveForceFlush(
+						persistenceContext.getEntry( old ) ).thenApply( v -> key );
+			}
+			else {
+				return failedFuture( new NonUniqueObjectException( id, persister.getEntityName() ) );
+			}
+		}
+
+		return CompletionStages.completedFuture( key );
 	}
 
 	/**

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveSession.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/ReactiveSession.java
@@ -13,6 +13,7 @@ import org.hibernate.LockMode;
 import org.hibernate.LockOptions;
 import org.hibernate.UnknownProfileException;
 import org.hibernate.collection.spi.PersistentCollection;
+import org.hibernate.engine.spi.EntityEntry;
 import org.hibernate.engine.spi.PersistenceContext;
 import org.hibernate.engine.spi.SessionImplementor;
 import org.hibernate.event.internal.MergeContext;
@@ -67,6 +68,8 @@ public interface ReactiveSession extends ReactiveQueryExecutor {
 	CompletionStage<Void> reactiveFlush();
 
 	CompletionStage<Void> reactiveAutoflush();
+
+	CompletionStage<Void> reactiveForceFlush(EntityEntry entry);
 
 	CompletionStage<Void> reactiveRefresh(Object entity, LockOptions lockMode);
 

--- a/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
+++ b/hibernate-reactive-core/src/main/java/org/hibernate/reactive/session/impl/ReactiveSessionImpl.java
@@ -898,6 +898,26 @@ public class ReactiveSessionImpl extends SessionImpl implements ReactiveSession,
 		return getHibernateFlushMode().lessThan( FlushMode.COMMIT ) ? voidFuture() : doFlush();
 	}
 
+	@Override
+	public CompletionStage<Void> reactiveForceFlush(EntityEntry entry) {
+		if ( log.isDebugEnabled() ) {
+			log.debugf(
+					"Flushing to force deletion of re-saved object: %s",
+					MessageHelper.infoString( entry.getPersister(), entry.getId(), getFactory() )
+			);
+		}
+
+		if ( getPersistenceContextInternal().getCascadeLevel() > 0 ) {
+			return CompletionStages.failedFuture( new ObjectDeletedException(
+					"deleted object would be re-saved by cascade (remove deleted object from associations)",
+					entry.getId(),
+					entry.getPersister().getEntityName()
+			) );
+		}
+		checkOpenOrWaitingForAutoClose();
+		return doFlush();
+	}
+
 	private CompletionStage<Void> doFlush() {
 		checkTransactionNeededForUpdateOperation( "no transaction is in progress" );
 		pulseTransactionCoordinator();

--- a/hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java
+++ b/hibernate-reactive-core/src/test/java/org/hibernate/reactive/MutinySessionTest.java
@@ -618,6 +618,25 @@ public class MutinySessionTest extends BaseReactiveTest {
 		context.assertEquals( expected.getName(), actual.getName() );
 	}
 
+	@Test
+	public void testForceFlushWithDelete(TestContext context) {
+		final GuineaPig pig1 = new GuineaPig( 111, "Aloi" );
+		final GuineaPig pig2 = new GuineaPig( 111, "Bloi" );
+
+		test( context, getMutinySessionFactory()
+				.withTransaction( session -> session
+						.persist( pig1 )
+						// remove pig1
+						.call( () -> session.remove( pig1 ) )
+						// persist pig2 with same ID as original pig1 to verify that reactive flush was performed
+						.call( () -> session.persist( pig2 ) )
+				).chain( () -> getMutinySessionFactory()
+						.withSession( s -> s.find( GuineaPig.class, pig2.getId() ) ) )
+						// check that found entity is "pig2"
+				.invoke( result -> assertThatPigsAreEqual( context, pig2, result ) )
+		);
+	}
+
 	@Entity(name="GuineaPig")
 	@Table(name="Pig")
 	public static class GuineaPig {

